cmd/migrate-curio/migrate.go

```diff
@@ -0,0 +1,557 @@
+			continue
+		}
+
+		// Skip if already migrated
+		var a, b, c bool
+		err = mdb.QueryRow(`SELECT DB, LID, Pipeline FROM Deals WHERE ID = ?`, deal.DealUuid.String()).Scan(&a, &b, &c)
+		if err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("deal: %s: failed to check migration status: %w", deal.DealUuid.String(), err)
+			}
+		}
+		if a && b && c {
+			llog.Infow("Skipped as this deal is already migrated")
+			continue
+		}
+
+		propJson, err := json.Marshal(deal.ClientDealProposal.Proposal)
+		if err != nil {
+			return fmt.Errorf("deal: %s: json.Marshal(piece.DealProposal): %s", deal.DealUuid.String(), err)
+		}
+
+		sigByte, err := deal.ClientDealProposal.ClientSignature.MarshalBinary()
+		if err != nil {
+			return fmt.Errorf("deal: %s: marshal client signature: %s", deal.DealUuid.String(), err)
+		}
+
+		prop := deal.ClientDealProposal.Proposal
+
+		sProp, err := deal.SignedProposalCid()
+		if err != nil {
+			return err
+		}
+
+		// de-serialize transport opaque token
+		tInfo := &transportTypes.HttpRequest{}
+		if err := json.Unmarshal(deal.Transfer.Params, tInfo); err != nil {
+			return fmt.Errorf("deal: %s: failed to de-serialize transport params bytes '%s': %s", deal.DealUuid.String(), string(deal.Transfer.Params), err)
+		}
+
+		headers, err := json.Marshal(tInfo.Headers)
+		if err != nil {
+			return fmt.Errorf("deal: %s: failed to marshal headers: %s", deal.DealUuid.String(), err)
+		}
+
+		// Add deal to HarmonyDB
+		if !a {
+			_, err = hdb.Exec(ctx, `INSERT INTO market_mk12_deals (uuid, sp_id, signed_proposal_cid, 
+                                proposal_signature, proposal, piece_cid, 
+                                piece_size, offline, verified, start_epoch, end_epoch, 
+                                client_peer_id, fast_retrieval, announce_to_ipni, url, url_headers, chain_deal_id, publish_cid, created_at) 
+				VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19)
+				ON CONFLICT (uuid) DO NOTHING`,
+				deal.DealUuid.String(), mid, sProp.String(), sigByte, propJson, prop.PieceCID.String(),
+				prop.PieceSize, deal.IsOffline, prop.VerifiedDeal, prop.StartEpoch, prop.EndEpoch, deal.ClientPeerID.String(),
+				deal.FastRetrieval, deal.AnnounceToIPNI, tInfo.URL, headers, int64(deal.ChainDealID), deal.PublishCID.String(), deal.CreatedAt)
+
+			if err != nil {
+				return fmt.Errorf("deal: %s: failed to add the deal to harmonyDB: %w", deal.DealUuid.String(), err)
+			}
+
+			// Mark deal added to harmonyDB
+			_, err = mdb.Exec(`INSERT INTO Deals (ID, DB, LID) VALUES (?, TRUE, FALSE) ON CONFLICT(ID) DO NOTHING`, deal.DealUuid.String())
+			if err != nil {
+				return fmt.Errorf("deal: %s: failed to mark deal migrated: %w", deal.DealUuid.String(), err)
+			}
+		}
+
+		if !b {
+			// Add LID details to pieceDeal in HarmonyDB
+			_, err = hdb.Exec(ctx, `SELECT process_piece_deal($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
```

Comment:
I guess this will just reimport the piece, redoing LID/IPNI will be a separate small reindexing pipeline, right?

Reply:
Yes. This is to ensure that there is no chance of missing metadata. Func process_piece_deal does not raise an exception if the piece is already present in the table. First time indexing task actually creates the indexes, we mark the piece as indexed and if another deal has the same piece then we skip the actual indexing next time around.